src/main/java/org/manifold/compiler/front/FunctionInvocationVertex.java

```diff
@@ -107,9 +113,74 @@ private void elaborateNodeInstantiation(Value function,
     // now remove this vertex from the graph
     getExpressionGraph().removeVertex(this);
   }
+
+  private void elaborateNonPrimitiveFunction(Value f,
+      ExpressionVertex input) throws Exception {
+    log.debug("function invocation is non-primitive elaboration");
+    FunctionValue function = (FunctionValue) f;
+    FunctionTypeValue signature = (FunctionTypeValue) f.getType();
+    TupleTypeValue inputType = (TupleTypeValue) signature.getInputType();
+    TupleTypeValue outputType = (TupleTypeValue) signature.getOutputType();
+    // calculate renaming map (body -> this.exprGraph)
+    Map<VariableReferenceVertex, VariableReferenceVertex> renamingMap =
+        new HashMap<>();
+    for (Entry<VariableIdentifier, VariableReferenceVertex> entry
+        : function.getBody().getVariableVertices().entrySet()) {
+      VariableReferenceVertex varTarget;
+      if (getExpressionGraph().containsVariable(entry.getKey())) {
+        // if the main graph contains this variable, don't rename; use existing
+        log.debug("preserving variable " + entry.getKey());
+        varTarget = getExpressionGraph().getVariableVertex(entry.getKey());
+      } else {
+        // otherwise, create a fresh variable in the main graph and map to that
+        String oldName = entry.getKey().getName();
+        Random rng = new Random();
+        while (true) {
+          // choose a random suffix and mangle
+          Integer i = rng.nextInt(Integer.MAX_VALUE);
+          String newName = oldName + "_" + i.toString();
+          List<String> newNames = new ArrayList<>();
+          newNames.add(newName);
+          VariableIdentifier newID = new VariableIdentifier(newNames);
+          if (!getExpressionGraph().containsVariable(newID)) {
```

Comment:
Ok, I see this being added here (WRT my comment in ExpressionGraph). I still think it makes more sense to keep it as VarId -> VarId, do the replacement and add the Vertex in the ExpressionGraph function. That way all cases of inserting nodes from this graph to the main graph occur in the same place, and that code is less confusing.